### PR TITLE
Support conditional breakpoints within debase.

### LINKED ISSUE #36: Conditional breakpoints always break.
When setting a conditional breakpoint, the resulting breakpoint is triggered every time the line is hit, not just when the specified condition is `true`.

In reviewing the code, I see that `ruby-debug-ide` accepts the `break ... if ...` command and passes it along to `debase`, which records the provided expression.  But then, neither `ruby-debug-ide` nor `debase` does anything with the expression.

Am I overlooking something?

### PR DESCRIPTION
**Pull request by seraku24**
This should allow `debase` to be able to perform conditional breakpoint evaluation on its own.  Fixes #36.

I have done basic acceptance testing against VS Code* and PuTTY connecting as IDEs.  @denofevil, I would be grateful if you could test these changes against RubyMine to confirm that I have not inadvertently broken compatibility.

(* `vscode-ruby` still has some pending work for enabling conditional breakpoints, so my testing is against a private build at the moment.)

And, thanks to @deivid-rodriguez and the `byebug` project, which proved to be a helpful resource.  `(:`

### CODE REVIEW

File: ext/debase_internals.c [hunk 1]
```diff
@@ -371,7 +373,7 @@ process_line_event(VALUE trace_point, void *data)
       context->stop_frame = -1;
     }
 
-    breakpoint = breakpoint_find(breakpoints, path, lineno);
+    breakpoint = breakpoint_find(breakpoints, path, lineno, binding);
```
Comment (denofevil) [inline ext/debase_internals.c]: I'd rather pass whole tracepoint object here for following reason:
binding calculation cost is super high, calling it on every line event will slow down debugging dramatically (more 2x) even if there are no conditional breakpoints. Thus getting binding should really be the latest operation before condition evaluation

Comment (seraku24) [inline ext/debase_internals.c]: Makes sense to defer, if there is a measurable performance hit.

But I do not see how the test functions will be able to pass in a `TracePoint` object.  Of course, we could cheat and have the tests pass in the `Binding` as they are doing, and have `process_line_event` pass in the `TracePoint` object.  (We're already abstracted behind the `VALUE` type, so this doesn't break typing.)  Then `check_breakpoint_expr` would verify the type of the `binding` parameter and, as needed, perform the actual `rb_tracearg_binding` call there.  But, that would be ugly mixing test-centric code with the main code.

An alternative is to use something like a promise.  (I presume Ruby has them; they seem to exist in some form or another in almost all languages.)  This results in nearly the same behavior as above, but it feels slightly less dirty.  Thoughts?

**Comment by deivid-rodriguez** [comment]
@seraku24 You're welcome, feel free to contribute back to `byebug` any improvements you think `byebug` could benefit from! :wink:

**Comment by seraku24** [comment]
@denofevil, let me know if this works for you.  The test code does takes a slightly different path than the real code, but this is the cleanest approach I found that does not overly complicate things.

**Comment by denofevil** [comment]
@seraku24, I've checked previous version of the code in ruby-debug-base and it seems that [position](https://github.com/ruby-debug/ruby-debug/blob/master/ext/ruby_debug.c#L813) and [expression](https://github.com/ruby-debug/ruby-debug/blob/master/ext/ruby_debug.c#L823) checks were separated, that probably will make tests more clean but also more unit-like.
In general LGTM, I can merge this one.

**Comment by denofevil** [comment]
@seraku24 what do you think, do you want to check previous implementation or we're merging this one?

**Comment by denofevil** [comment]
It should also fix #11

**Comment by seraku24** [comment]
@denofevil, I did not realize you had been waiting on me.  I am very sorry about that.

I just reviewed the files you linked.  At this point, I do not believe the differences from what we have here should matter in the grand scheme of things.

**Comment by denofevil** [comment]
@seraku24 no problems, let me merge this then and update the gem

**Comment by denofevil** [comment]
Pushed debase 0.2.2.beta10 to rubygems.org.

**Pull request merged**